file_util.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2021 Snowflake Computing Inc. All right reserved.
+
+package gosnowflake
+
+//lint:file-ignore U1000 TODO SNOW-29352
+
+import (
+	"bytes"
+	"compress/gzip"
+	"crypto/sha256"
+	"encoding/base64"
+	"io"
+	"io/ioutil"
+	usr "os/user"
+	"path/filepath"
+	"strings"
+)
+
+type snowflakeFileUtil struct {
+}
+
+func (util *snowflakeFileUtil) compressFileWithGzipFromStream(srcStream **bytes.Buffer) (*bytes.Buffer, int) {
+	buf := getByteBufferContent(srcStream)
+	var c bytes.Buffer
+	w := gzip.NewWriter(&c)
+	w.Write(*buf) // write buf to gzip writer
+	w.Close()
+	return &c, c.Len()
+}
+
+func (util *snowflakeFileUtil) getDigestAndSize(src **bytes.Buffer) (string, int64) {
+	m := sha256.New()
+	buf := getByteBufferContent(src)
+	m.Write(*buf)
+	return base64.StdEncoding.EncodeToString(m.Sum(nil)), int64((*src).Len())
+}
+
+func (util *snowflakeFileUtil) getDigestAndSizeForStream(stream **bytes.Buffer) (string, int64) {
+	return util.getDigestAndSize(stream)
+}
+
+func (util *snowflakeFileUtil) getDigestAndSizeForFile(fileName string) (string, int64) {
+	src, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		panic(err)
+	}
+	buf := bytes.NewBuffer(src)
+	return util.getDigestAndSize(&buf)
+}
+
+// file metadata for PUT/GET
+type fileMetadata struct {
```

Comment:
I know this is how other driver implemented fileMetadata, but some items are of 'the file to transfer', some are related to 'the transferring', and some are related to the 'transfer worker'. While reorg the code could be hard, I would suggest to check this point: when there is 'sfa', do we still need six 'callback's in fileMeta?

Reply:
I see. Do you have a suggestion?

Reply:
I will look into the re-org when I start implementing the S3 PUT

Reply:
I think you could remove callbacks in first commit. They are not used anywhere in this commit. And in next commit you could use sfa.callback directly.